File: 03-python-fastapi-websocket/local_mic.py
"""local_mic.py — talk to the agent from your terminal (server-to-server).

    python local_mic.py

Captures the default microphone with pyaudio at 24 kHz, streams it to Silk,
and plays the agent's reply on the default output. Shows barge-in: when Silk
sends `interruption`, the playback queue is dropped immediately.

macOS, Windows and ALSA's `default` device all resample transparently, so
opening the device at 24 kHz works almost everywhere. If your device refuses,
resample to 24 kHz before sending (see docs/audio-resampling.md).
"""

from __future__ import annotations

import asyncio
import os
import sys
import threading

import pyaudio
from dotenv import load_dotenv

from silk_client import FRAME_SAMPLES, SAMPLE_RATE, SilkCallError, SilkRealtimeSession, register_call

load_dotenv()
API_KEY = os.environ["RUMIK_API_KEY"]
AGENT_ID = os.environ["RUMIK_AGENT_ID"]


async def main() -> None:
    reg = await register_call(API_KEY, AGENT_ID)           # {access_token, expires_in, sample_rate}
    session = SilkRealtimeSession(reg["access_token"])
    created = await session.connect()                        # waits for session.created
    print(f"call {created['call_id']} · {created['sample_rate']} Hz · Ctrl+C to hang up\n")

    loop = asyncio.get_running_loop()
    pa = pyaudio.PyAudio()

    # ---- microphone → asyncio queue (pyaudio callbacks run on their own thread)
    mic_q: asyncio.Queue[bytes] = asyncio.Queue()

    def mic_cb(in_data, _frames, _time, _status):
        loop.call_soon_threadsafe(mic_q.put_nowait, in_data)
        return (None, pyaudio.paContinue)

    # ---- agent audio → playback ring buffer, drained by the output callback
    play_buf = bytearray()
    play_lock = threading.Lock()

    def spk_cb(_in, frame_count, _time, _status):
        need = frame_count * 2
        with play_lock:
            chunk = bytes(play_buf[:need])
            del play_buf[:need]
        if len(chunk) < need:                                # underrun → pad with silence
            chunk += b"\x00" * (need - len(chunk))
        return (chunk, pyaudio.paContinue)

    mic = pa.open(format=pyaudio.paInt16, channels=1, rate=SAMPLE_RATE, input=True,
                  frames_per_buffer=FRAME_SAMPLES, stream_callback=mic_cb)
    spk = pa.open(format=pyaudio.paInt16, channels=1, rate=SAMPLE_RATE, output=True,
                  frames_per_buffer=FRAME_SAMPLES, stream_callback=spk_cb)

    state = {"agent_speaking": False}

    async def on_audio(pcm: bytes) -> None:
        with play_lock:
            play_buf.extend(pcm)

    async def on_event(ev: dict) -> None:
        t = ev["type"]
        if t == "agent_start_talking":
            state["agent_speaking"] = True
            print("\033[91m● agent speaking\033[0m")                    # red
        elif t == "agent_stop_talking":
            state["agent_speaking"] = False
            print("\033[90m○ listening\033[0m")                         # grey
        elif t == "interruption":
            with play_lock:
                dropped = len(play_buf)
                play_buf.clear()                                        # ← barge-in: flush
            print(f"\033[92m● you interrupted — dropped {dropped / 48:.0f} ms of queued audio\033[0m")  # green
        elif t == "transcript":
            print(f"   {ev['role']:>9}: {ev['text']}")
        elif t == "session.closed":
            print(f"\ncall ended: {ev['reason']}")
        elif t == "error":
            print(f"\nerror {ev.get('code')}: {ev.get('message')}", file=sys.stderr)

    session.on_audio = on_audio
    session.on_event = on_event

    async def pump_mic() -> None:
        while True:
            await session.send_audio(await mic_q.get())

    mic_task = asyncio.create_task(pump_mic())
    try:
        await session.run()                                             # until session.closed
    except asyncio.CancelledError:
        pass
    finally:
        mic_task.cancel()
        await session.close()
        mic.close(); spk.close(); pa.terminate()


if __name__ == "__main__":
    try:
        asyncio.run(main())
    except KeyboardInterrupt:
        print("\nhung up")
    except SilkCallError as e:
        print(f"could not start call: {e}", file=sys.stderr)
        sys.exit(1)

File: 04-nodejs-websocket/local-mic.js
// local-mic.js — talk to the agent from your terminal (server-to-server).
//
//   npm run mic
//
// Records the microphone with node-record-lpcm16 (which shells out to sox /
// rec / arecord), resamples to 24 kHz if needed, streams to Silk, and plays
// the reply through the `speaker` package. Shows barge-in: on `interruption`
// the playback queue is dropped immediately.
//
// Install the recorder binary first: macOS `brew install sox`, Debian
// `apt install sox alsa-utils`.

import "dotenv/config";
import recorder from "node-record-lpcm16";
import Speaker from "speaker";
import { FRAME_BYTES, SAMPLE_RATE, SilkRealtimeSocket, registerCall } from "./lib/silk-socket.js";
import { Resampler, bufferToInt16, int16ToBuffer } from "./lib/resampler.js";

const { RUMIK_API_KEY, RUMIK_AGENT_ID } = process.env;
const MIC_RATE = Number(process.env.MIC_RATE ?? SAMPLE_RATE);

const red = (s) => `\x1b[91m${s}\x1b[0m`, green = (s) => `\x1b[92m${s}\x1b[0m`, grey = (s) => `\x1b[90m${s}\x1b[0m`;

/**
 * Paced playback: we keep our own queue and feed the speaker exactly one 20 ms
 * frame every 20 ms. That keeps the device's internal buffer shallow, so a
 * flush() on interruption is heard almost immediately.
 */
class PacedPlayer {
  constructor() {
    this.speaker = new Speaker({ channels: 1, bitDepth: 16, sampleRate: SAMPLE_RATE, signed: true });
    this.queue = Buffer.alloc(0);
    this.silence = Buffer.alloc(FRAME_BYTES);
    this.timer = setInterval(() => this.tick(), 20);
  }
  push(pcm) { this.queue = Buffer.concat([this.queue, pcm]); }
  flush() { const dropped = this.queue.length; this.queue = Buffer.alloc(0); return dropped; }
  tick() {
    if (this.queue.length >= FRAME_BYTES) {
      this.speaker.write(this.queue.subarray(0, FRAME_BYTES));
      this.queue = this.queue.subarray(FRAME_BYTES);
    } else {
      this.speaker.write(this.silence); // keep the stream continuous
    }
  }
  stop() { clearInterval(this.timer); this.speaker.end(); }
}

async function main() {
  const reg = await registerCall({ apiKey: RUMIK_API_KEY, agentId: RUMIK_AGENT_ID });
  const call = new SilkRealtimeSocket(reg.access_token);
  const player = new PacedPlayer();
  const resampler = MIC_RATE === SAMPLE_RATE ? null : new Resampler(MIC_RATE, SAMPLE_RATE);

  call.on("audio", (pcm) => player.push(pcm));
  call.on("agent_start_talking", () => console.log(red("● agent speaking")));
  call.on("agent_stop_talking", () => console.log(grey("○ listening")));
  call.on("interruption", () => {
    const dropped = player.flush(); // ← barge-in: drop everything queued
    console.log(green(`● you interrupted — dropped ${Math.round(dropped / 48)} ms of queued audio`));
  });
  call.on("transcript", ({ role, text }) => console.log(`   ${role.padStart(9)}: ${text}`));
  call.on("silk_error", (e) => console.error(`error ${e.code}: ${e.message}`));
  call.on("session.closed", ({ reason }) => console.log(`\ncall ended: ${reason}`));
  call.on("close", () => { player.stop(); rec.stop(); process.exit(0); });

  const created = await call.connect(); // waits for session.created
  console.log(`call ${created.call_id} · ${created.sample_rate} Hz · mic at ${MIC_RATE} Hz · Ctrl+C to hang up\n`);

  // microphone → (resample) → Silk
  const rec = recorder.record({ sampleRate: MIC_RATE, channels: 1, audioType: "raw", threshold: 0 });
  rec.stream().on("data", (chunk) => {
    const pcm = resampler ? int16ToBuffer(resampler.process(bufferToInt16(chunk))) : chunk;
    call.sendAudio(pcm);
  });

  process.on("SIGINT", () => { console.log("\nhanging up…"); call.close(); });
}

main().catch((err) => { console.error("could not start call:", err.message); process.exit(1); });
